Bunny Bcbio support

To continue discussion from #92

Thanks for the latest bunny release, I'm excited to keep moving bcbio CWL support forward. I tested out with my small self contained test case (https://github.com/bcbio/test_bcbio_cwl) which you can run with the `run_bunny.sh` shell script. I'm running into a different null pointer error:
```
[2017-02-15 05:59:33.874] [ERROR] EventProcessor failed to process event OutputUpdateEvent [jobId=root.prep_samples.2, contextId=6509ef2d-1198-499a-a31d-80137c12560c, portId=config__algorithm__variant_regions, value=FileValue [size=65, path=/home/chapmanb/drive/work/cwl/test_bcbio_cwl/bunny_work/main-run_info-cwl-20170215055916194/root/prep_samples/2/bedprep/variant_regions-bam.bed, location=/home/chapmanb/drive/work/cwl/test_bcbio_cwl/bunny_work/main-run_info-cwl-20170215055916194/root/prep_samples/2/bedprep/variant_regions-bam.bed, checksum=sha1$02cb0626caa3c5d7db51b0ff867553d7ff5422d6, secondaryFiles=[], properties={sbg:metadata=null}], fromScatter=false, numberOfScattered=null].
java.lang.NullPointerException: null
        at org.rabix.engine.model.VariableRecord.expand(VariableRecord.java:101) ~[rabix-backend-local.jar:na]
        at org.rabix.engine.model.VariableRecord.addValue(VariableRecord.java:80) ~[rabix-backend-local.jar:na]
        at org.rabix.engine.processor.handler.impl.InputEventHandler.handle(InputEventHandler.java:73) ~[rabix-backend-local.jar:na]
        at org.rabix.engine.processor.handler.impl.InputEventHandler.handle(InputEventHandler.java:30) ~[rabix-backend-local.jar:na]
        at org.rabix.engine.processor.dispatcher.impl.SyncEventDispatcher.send(SyncEventDispatcher.java:21) ~[rabix-backend-local.jar:na]
        at org.rabix.engine.processor.impl.EventProcessorImpl.send(EventProcessorImpl.java:132) ~[rabix-backend-local.jar:na]
        at org.rabix.engine.processor.impl.MultiEventProcessorImpl.send(MultiEventProcessorImpl.java:55) ~[rabix-backend-local.jar:na]
        at sun.reflect.GeneratedMethodAccessor10.invoke(Unknown Source) ~[na:na]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_121]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_121]
        at com.google.inject.internal.DelegatingInvocationHandler.invoke(DelegatingInvocationHandler.java:50) ~[rabix-backend-local.jar:na]
        at com.sun.proxy.$Proxy37.send(Unknown Source) ~[na:na]
        at org.rabix.engine.processor.handler.impl.OutputEventHandler.handle(OutputEventHandler.java:198) ~[rabix-backend-local.jar:na]
        at org.rabix.engine.processor.handler.impl.OutputEventHandler.handle(OutputEventHandler.java:43) ~[rabix-backend-local.jar:na]
        at org.rabix.engine.processor.impl.EventProcessorImpl$1.run(EventProcessorImpl.java:79) ~[rabix-backend-local.jar:na]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_121]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_121]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_121]
```
From the logs all of the launched tasks appear to complete but then there is some issue merging the outputs. Happy to work more on this from the bcbio side with any details about what triggers this. Thank you again for all the work.

--- a/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/CWLCommandLineBuilder.java
+++ b/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/CWLCommandLineBuilder.java
@@ -33,23 +33,12 @@ import org.slf4j.LoggerFactory;
 import com.google.common.base.Function;
 import com.google.common.base.Joiner;
 import com.google.common.collect.Lists;
-import com.google.common.escape.Escaper;
-import com.google.common.escape.Escapers;
 
 public class CWLCommandLineBuilder implements ProtocolCommandLineBuilder {
 
   private final static Logger logger = LoggerFactory.getLogger(CWLCommandLineBuilder.class);
   
   public final static String SHELL_QUOTE_KEY = "shellQuote";
-  public final static String SHELL_QUOTE_POSITION = "position";
-  public final static String SHELL_QUOTE_VALUE_KEY = "valueFrom";
-  
-  public static final Escaper SHELL_ESCAPE;
-  static {
-      final Escapers.Builder builder = Escapers.builder();
-      builder.addEscape('\'', "'\"'\"'");
-      SHELL_ESCAPE = builder.build();
-  }
   
   @Override
   public CommandLine buildCommandLineObject(Job job, File workingDir, FilePathMapper filePathMapper) throws BindingException {
@@ -115,39 +104,12 @@ public class CWLCommandLineBuilder implements ProtocolCommandLineBuilder {
   }
   
   /**
-   * Is shellQuote enabled
-   */
-  private boolean isShellQuote(Object input) {
-    if (input == null) {
-      return false;
-    }
-    if (input instanceof Map<?,?>) {
-      return CWLBeanHelper.getValue(SHELL_QUOTE_KEY, input) != null;
-    }
-    return false;
-  }
-
-  /**
    * Get shellQuote flag 
    */
   private boolean getShellQuote(Object input) {
-    return CWLBeanHelper.getValue(SHELL_QUOTE_KEY, input);
+    return CWLBeanHelper.getValue(SHELL_QUOTE_KEY, input, true);
   }
   
-  /**
-   * Get shellQuote value 
-   */
-  private Object getShellQuoteValue(Object input) {
-    return CWLBeanHelper.getValue(SHELL_QUOTE_VALUE_KEY, input);
-  }
-  
-  /**
-   * Get shellQuote value 
-   */
-  private int getShellQuotePosition(Object input) {
-    return CWLBeanHelper.getValue(SHELL_QUOTE_POSITION, input, 0);
-  }
-
   /**
    * Build command line arguments
    */
@@ -167,25 +129,16 @@ public class CWLCommandLineBuilder implements ProtocolCommandLineBuilder {
 
       if (commandLineTool.hasArguments()) {
         for (int i = 0; i < commandLineTool.getArguments().size(); i++) {
-          int position = 0;
-          boolean shellQuote = false;
+
           Object argBinding = commandLineTool.getArguments().get(i);
-          if (isShellQuote(argBinding)) {
-            position = getShellQuotePosition(argBinding);
-            shellQuote = getShellQuote(argBinding);
-            argBinding = getShellQuoteValue(argBinding);
+          Object argValue;
+          if (argBinding instanceof Map<?,?>) {
+            argValue = CWLBeanHelper.getValue(CWLCommandLineTool.KEY_ARGUMENT_VALUE, argBinding);
+          } else {
+            argValue = argBinding;
+            argBinding = Collections.singletonMap(CWLCommandLineTool.KEY_ARGUMENT_VALUE, argValue);
           }
-          if (argBinding instanceof String) {
-            Object arg = CWLExpressionResolver.resolve(argBinding, job, null);
-            if (shellQuote) {
-              arg = EncodingHelper.shellQuote(arg);
-            }
-            CWLCommandLinePart commandLinePart = new CWLCommandLinePart.Builder(position, false).part(arg).keyValue("").build();
-            commandLinePart.setArgsArrayOrder(i);
-            commandLineParts.add(commandLinePart);
-            continue;
-          }
-          Object argValue = commandLineTool.getArgument(job, argBinding);
+
           Map<String, Object> emptySchema = new HashMap<>();
           CWLCommandLinePart commandLinePart = buildCommandLinePart(job, null, argBinding, argValue, emptySchema, null);
           if (commandLinePart != null) {
@@ -260,11 +213,7 @@ public class CWLCommandLineBuilder implements ProtocolCommandLineBuilder {
     String itemSeparator = CWLBindingHelper.getItemSeparator(inputBinding);
     String keyValue = inputPort != null ? inputPort.getId() : "";
 
-    boolean shellQuote = false;
-    if (isShellQuote(value)) {
-      shellQuote = getShellQuote(value);
-      value = getShellQuoteValue(value);
-    }
+    boolean shellQuote = getShellQuote(inputBinding);
     
     Object valueFrom = CWLBindingHelper.getValueFrom(inputBinding);
     if (valueFrom != null) {
@@ -369,7 +318,7 @@ public class CWLCommandLineBuilder implements ProtocolCommandLineBuilder {
     if (shellQuote) {
       value = EncodingHelper.shellQuote(value);
     }
-    
+
     if (prefix == null) {
       return new CWLCommandLinePart.Builder(position, isFile).keyValue(keyValue).part(value).build();
     }

--- a/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/bean/CWLCommandLineTool.java
+++ b/rabix-bindings-cwl/src/main/java/org/rabix/bindings/cwl/bean/CWLCommandLineTool.java
@@ -100,17 +100,6 @@ public class CWLCommandLineTool extends CWLJobApp {
     return arguments;
   }
 
-  @JsonIgnore
-  public Object getArgument(CWLJob job, Object binding) throws CWLExpressionException {
-    if (binding instanceof Map<?, ?>) {
-      Object value = ((Map<?, ?>) binding).get(KEY_ARGUMENT_VALUE);
-      if (value != null) {
-        return CWLExpressionResolver.resolve(value, job, null);
-      }
-    }
-    return CWLExpressionResolver.resolve(binding, job, null);
-  }
-
   public static String generateRandomStdoutGlob() {
     return RANDOM_STDOUT_PREFIX + UUID.randomUUID().toString();
   }

--- a/rabix-common/src/main/java/org/rabix/common/helper/EncodingHelper.java
+++ b/rabix-common/src/main/java/org/rabix/common/helper/EncodingHelper.java
@@ -1,12 +1,29 @@
 package org.rabix.common.helper;
 
 import java.nio.charset.Charset;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
+import com.google.common.escape.Escaper;
+import com.google.common.escape.Escapers;
 import org.apache.commons.codec.binary.Base64;
 
 public class EncodingHelper {
 
   public static Charset DEFAULT_ENCODING = Charset.forName("UTF-8");
+
+  public static final String SHELL_CHARS_NEEDS_QUOTING = "|&;<>()$`\\ \"'\t\r\n*?[#~";
+  public static final Pattern NEEDS_QUOTING_PATTERN = Pattern.compile("[" + Pattern.quote(SHELL_CHARS_NEEDS_QUOTING) + "]");
+
+  public static final Escaper SHELL_ESCAPE;
+  static {
+    final Escapers.Builder builder = Escapers.builder();
+    builder.addEscape('$', "\\$");
+    builder.addEscape('`', "\\`");
+    builder.addEscape('\\', "\\\\");
+    builder.addEscape('"', "\\\"");
+    SHELL_ESCAPE = builder.build();
+  }
   
   public static String encodeBase64(String data) {
     return Base64.encodeBase64String(data.getBytes(DEFAULT_ENCODING));
@@ -25,11 +42,10 @@ public class EncodingHelper {
       return argument;
     }
     String argumentStr = (String) argument;
-    if (argumentStr.contains("'")) {
-      argumentStr = argumentStr.replace("'", "\\'");
-      return "'" + argumentStr + "'";
+    if (!NEEDS_QUOTING_PATTERN.matcher(argumentStr).find()) {
+      return argumentStr;
     }
-    return argument;
+    return '"' + SHELL_ESCAPE.escape(argumentStr) + '"';
   }
   
 }

